flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducerBaseTest.java

```diff
@@ -88,195 +87,296 @@ public void testKeyValueDeserializersSetIfMissing() throws Exception {
-					for (Callback c: pending) {
-						c.onCompletion(null, null);
-					}
-					pending.clear();
-				} catch(Throwable t) {
-					runnableError.f0 = t;
+					// this should block at first, since there are still two pending records that needs to be flushed
+					testHarness.snapshot(123L, 123L);
+				} catch (Exception e) {
+					asyncError.f0 = e;
 				}
 			}
-		};
-		Thread threadB = new Thread(confirmer);
-		threadB.start();
+		});
+		snapshotThread.start();
 
-		// this should block:
-		testHarness.snapshot(0, 0);
+		// let the 2nd message fail with an async exception
+		producer.getPendingCallbacks().get(1).onCompletion(null, new Exception("artificial async failure for 2nd message"));
+		producer.getPendingCallbacks().get(2).onCompletion(null, null);
 
-		synchronized (threadA) {
-			threadA.notifyAll(); // just in case, to let the test fail faster
-		}
-		Assert.assertEquals(0, producer.getProducerInstance().getPending().size());
-		Deadline deadline = FiniteDuration.apply(5, "s").fromNow();
-		while (deadline.hasTimeLeft() && threadB.isAlive()) {
-			threadB.join(500);
+		snapshotThread.join();
+
+		// the snapshot should have failed with the async exception
+		Assert.assertTrue(asyncError.f0 != null && asyncError.f0.getCause().getMessage().contains("artificial async failure for 2nd message"));
+	}
+
+	/**
+	 * Test ensuring that the producer is not dropping buffered records;
+	 * we set a timeout because the test will not finish if the logic is broken
+	 */
+	@Test(timeout=10000)
+	public void testAtLeastOnceProducer() throws Throwable {
+
+		KafkaProducer mockProducer = mock(KafkaProducer.class);
+		final DummyFlinkKafkaProducer<String> producer = new DummyFlinkKafkaProducer<>(
+			FakeStandardProducerConfig.get(), null, mockProducer);
+		producer.setFlushOnCheckpoint(true);
+
+		final OneInputStreamOperatorTestHarness<String, Object> testHarness =
+			new OneInputStreamOperatorTestHarness<>(new StreamSink(producer));
+
+		testHarness.open();
+
+		testHarness.processElement(new StreamRecord<>("msg-1"));
+		testHarness.processElement(new StreamRecord<>("msg-2"));
+		testHarness.processElement(new StreamRecord<>("msg-3"));
+
+		verify(mockProducer, times(3)).send(any(ProducerRecord.class), any(Callback.class));
+		Assert.assertEquals(3, producer.getPendingSize());
+
+		// start a thread to perform checkpointing
+		final Tuple1<Throwable> runnableError = new Tuple1<>(null);
+		final OneShotLatch snapshotReturnedLatch = new OneShotLatch();
+
+		Thread snapshotThread = new Thread(new Runnable() {
+			@Override
+			public void run() {
+				try {
+					// this should block until all records are flushed
+					testHarness.snapshot(123L, 123L);
+				} catch (Throwable e) {
+					runnableError.f0 = e;
+				} finally {
+					snapshotReturnedLatch.trigger();
+				}
+			}
+		});
+		snapshotThread.start();
+
+		// being extra safe that the snapshot is correctly blocked
+		try {
+			snapshotReturnedLatch.await(3, TimeUnit.SECONDS);
```

Comment:
Maybe you could override the `DummyFlinkKafkaProducer#flush` method to insert some latches to see when you enter and when the method is done. Then you could wait on the first latch and check with the latter whether the method has completed.